pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -1123,3 +1123,23 @@ func (os *OpenStack) DeleteVolume(volumeName string) error {
 	}
 	return err
 }
+
+// Get device path of attached volume to the compute running kubelet
+func (os *OpenStack) GetAttachmentDiskPath(instanceID string, diskName string) (string, error) {
+	disk, err := os.getVolume(diskName)
+	if err != nil {
+		return "", err
+	}
+	if len(disk.Attachments) > 0 && disk.Attachments[0]["server_id"] != nil {
+		if instanceID == disk.Attachments[0]["server_id"] {
+			// Attachment[0]["device"] points to the device path
+			// see http://developer.openstack.org/api-ref-blockstorage-v1.html
+			return disk.Attachments[0]["device"].(string), nil
```

Comment:
here is what i've found.
- `nova volume-attach instance_name volume_id` (i.e. not specifying device), then the device path from `nova volume-attachments instance_name` matches that in the VM
- `nova volume-attach instance_name volume_id /dev/vdz` (`vdz` is a random device path), then the device path from `nova volume-attachments instance_name` doesn't match that in the VM

In openstack provider code, the volume attach doesn't take a device name as input, so it works for me.